### DISCUSSION TITLE
Security

### DISCUSSION THREAD

**Discussion by taaib** [m-ld/m-ld-spec#51] [Discussion]
How does m-ld ensure only ‘authorised’ clones join the cluster and that data from ‘bad actors’ is rejected?

**Comment by taaib** [comment]
Is the data that m-ld stores tied (in any way) to the user that created that data?

**Reply by gsvarovsky** [reply]
At the moment, data is not natively or automatically tied to participants at the level of **m-ld**. It is possible to achieve this kind of association in the data itself, just by having identities for users and by linking those users to the data they create with a JSON property.

There are two related future developments that are relevant to this:
1. The ability to tie transactions (and their resultant data) to user identity using a cryptographic signature, inside of **m-ld**. This will help clones and apps to reject transactions from attackers while preventing those attackers from being able to pretend they're someone else.
2. The ability to encrypt data with user secrets, such that it is possible to control data visibility and reject unauthorised edits.

Both of these enhancements are being researched at the moment. Let us know if they're of interest and what your specific requirements might be!

**Comment by gsvarovsky** [comment]
Hi @taaib, thanks for your questions.

In my reckoning, there are a few possible categories of 'bad actors'.

### Unauthorised users
In principle, a **m-ld** clone accepts its 'network' connection from the app, when it's initialised. (By 'network' I really mean whatever messaging service is being used – at the moment there are adapters for MQTT and Ably.) So, the app is able to secure that connection in any way that makes sense. Often, this will be by authenticating the local user and only allowing the connection if the user is authorised. A simple way to do this is to have user-specific accounts on the messaging service itself.

### Malware
A user might have legitimate authorisation, but their device has been compromised with malware, and so the application they _think_ they are logging into is not the real one. This scenario is not really any different for **m-ld** than for any other technology. You need to keep your anti-virus up to date, and only accept signed apps...

### Insiders
A user might have legitimate authorisation, but nefarious purposes. This can be hard to detect, and doing so is the responsibility of other users, maybe augmented by some kind of AI in the system architecture. Once detected, obviously the user's authorisation can be rescinded. However, what to do about the data they might have already corrupted? Of course if the data can be identified it can be manually fixed up. Another alternative is to have a 'backup' clone running on a server, which pushes all local transactions to a database. It would then be possible to create a new replica domain from a backup of this data. Again, in the end this is not special for **m-ld**, it's the same kind of thing that must be considered in all distributed data architectures.

There's lots more on https://m-ld.org/doc/#security. Do let us know about your use-case and what threats you're concerned about!
